semantics/executable-spec/src/Data/MemoBytes.hs

```diff
@@ -37,18 +40,24 @@ import Cardano.Binary
   )
 import Data.ByteString.Short (ShortByteString, fromShort, toShort)
 import qualified Data.ByteString.Lazy as Lazy
-import Data.ByteString.Lazy (toStrict)
+import Data.ByteString.Lazy (toStrict,fromStrict)
 import Data.Typeable
 import Data.Coders(runE, Encode, encode,)
 import Codec.CBOR.Write (toLazyByteString)
 import GHC.Generics (Generic)
-import NoThunks.Class (AllowThunksIn (..), NoThunks (..))
+import NoThunks.Class (NoThunks (..))
 import Prelude hiding (span)
+import GHC.TypeLits(Symbol)
+import GHC.Records
+import Codec.CBOR.Read(DeserialiseFailure,deserialiseFromBytes)
 
 -- ========================================================================
 
-data MemoBytes t = Memo {memotype :: !t, memobytes :: {-# UNPACK #-} !ShortByteString}
-  deriving (NoThunks) via AllowThunksIn '["memobytes"] (MemoBytes t)
+data MemoBytes t = Memo !t {-# UNPACK #-} !ShortByteString
+
+instance NoThunks t => NoThunks(MemoBytes t) where
+  noThunks ctxt (Memo x _) = noThunks ("Memo":ctxt) x
+  -- We deliberately allow thunks in the bytes
```

Comment:
What's the reason for this change? It was so pretty :cry: 

Moreover this manual instance is *wrong*. The [documentation of `NoThunks`](https://hackage.haskell.org/package/nothunks-0.1.1.0/docs/NoThunks-Class.html#v:noThunks) says:

> The default implementation of `noThunks` checks that the argument is in WHNF, and if so, adds the type into the context (using `showTypeOf`), and calls `wNoThunks`. See `ThunkInfo` for a detailed discussion of the type context.

So the method you should be overriding is `wNoThunks`, not `noThunks`. This is exactly why we provide the deriving-via wrappers like `AllowThunksIn` and `OnlyCheckWhnf`.

Reply:
Thanks for pointing this out, the original need for this was some HasRecord instances did not let MemoBytes have its own record fields. But since we no longer need this, I have rolled it back. Also thanks reference to the video about NoThunks. Now I know why what we did there was bad.

Reply:
Great, looks better now :+1: